### ISSUE TITLE
Error in Serialization of models.

### ISSUE THREAD

**Issue by kartikdutt18** [mlpack/mlpack#2561]
Hey there everyone,
There might be some error in serialization in Convolution or BatchNorm i.e. the saved model doesn't generate the same prediction when loaded again and used for prediction. Here is a simple file for testing.
```
#include <mlpack/core.hpp>
#include <mlpack/core/data/split_data.hpp>

#include <mlpack/methods/ann/layer/layer.hpp>
#include <mlpack/methods/ann/ffn.hpp>

#include <ensmallen.hpp>

using namespace mlpack;
using namespace mlpack::ann;

using namespace arma;
using namespace std;

using namespace ens;

arma::Row<size_t> getLabels(arma::mat predOut)
{
  arma::Row<size_t> predLabels(predOut.n_cols);
  for (arma::uword i = 0; i < predOut.n_cols; ++i)
  {
    predLabels(i) = predOut.col(i).index_max() + 1;
  }
  return predLabels;
}

int main()
{
  constexpr double RATIO = 0.1;

  // Allow at max 5 iterations.
  constexpr int MAX_ITERATIONS = 5;

  // Step size of the optimizer.
  constexpr double STEP_SIZE = 1.2e-3;

  // Number of data points in each iteration of SGD.
  constexpr int BATCH_SIZE = 50;

  cout << "Reading data ..." << endl;

  mat tempDataset;


  data::Load("../data/train.csv", tempDataset, true);


  mat dataset =
      tempDataset.submat(0, 1, tempDataset.n_rows - 1, tempDataset.n_cols - 1);
  mat train, valid;
  data::Split(dataset, train, valid, RATIO);
  const mat trainX = train.submat(1, 0, train.n_rows - 1, train.n_cols - 1) / 255;
  const mat validX = valid.submat(1, 0, valid.n_rows - 1, valid.n_cols - 1) / 255;


  const mat trainY = train.row(0) + 1;
  const mat validY = valid.row(0) + 1;
  FFN<NegativeLogLikelihood<>, RandomInitialization> model;

  model.Add<Convolution<>>(1,  // Number of input activation maps.
                           6,  // Number of output activation maps.
                           5,  // Filter width.
                           5,  // Filter height.
                           1,  // Stride along width.
                           1,  // Stride along height.
                           0,  // Padding width.
                           0,  // Padding height.
                           28, // Input width.
                           28  // Input height.
  );

  // Add BatchNorm layer.
  model.Add<BatchNorm<>>(6, 1e-8, false);

  // Add first ReLU.
  model.Add<LeakyReLU<>>();

  // Add first pooling layer. Pools over 2x2 fields in the input.
  model.Add<MaxPooling<>>(2, // Width of field.
                          2, // Height of field.
                          2, // Stride along width.
                          2, // Stride along height.
                          true);

  // Add the second convolution layer.
  model.Add<Convolution<>>(6,  // Number of input activation maps.
                           16, // Number of output activation maps.
                           5,  // Filter width.
                           5,  // Filter height.
                           1,  // Stride along width.
                           1,  // Stride along height.
                           0,  // Padding width.
                           0,  // Padding height.
                           12, // Input width.
                           12  // Input height.
  );

  // Add BatchNorm layer.
  model.Add<BatchNorm<>>(16, 1e-8, false);

  // Add the second ReLU.
  model.Add<LeakyReLU<>>();

  // Add the second pooling layer.
  model.Add<MaxPooling<>>(2, 2, 2, 2, true);

  // Add the final dense layer.
  model.Add<Linear<>>(16 * 4 * 4, 10);
  model.Add<LogSoftMax<>>();

  cout << "Start training ..." << endl;

  // Set parameters for the Adam optimizer.
  ens::Adam optimizer(
      STEP_SIZE,  // Step size of the optimizer.
      BATCH_SIZE, // Batch size. Number of data points that are used in each
                  // iteration.
      0.9,        // Exponential decay rate for the first moment estimates.
      0.999, // Exponential decay rate for the weighted infinity norm estimates.
      1e-8,  // Value used to initialise the mean squared gradient parameter.
      MAX_ITERATIONS * trainY.n_cols, // Max number of iterations.
      1e-8,           // Tolerance.
      true);

  model.Train(trainX,
              trainY,
              optimizer,
              ens::PrintLoss(),
              ens::ProgressBar(),
              // Stop the training using Early Stop at min loss.
              ens::EarlyStopAtMinLoss(
                  [&](const arma::mat& /* param */)
                  {
                    double validationLoss = model.Evaluate(validX, validY);
                    std::cout << "Validation loss: " << validationLoss
                        << "." << std::endl;
                    return validationLoss;
                  }));

  // Matrix to store the predictions on train and validation datasets.
  mat predOut;
  // Get predictions on training data points.
  model.Predict(trainX, predOut);
  // Calculate accuracy on training data points.
  arma::Row<size_t> predLabels = getLabels(predOut);
  double trainAccuracy =
      arma::accu(predLabels == trainY) / (double) trainY.n_elem * 100;
  // Get predictions on validating data points.
  model.Predict(validX, predOut);
  // Calculate accuracy on validating data points.
  predLabels = getLabels(predOut);
  double validAccuracy =
      arma::accu(predLabels == validY) / (double) validY.n_elem * 100;

  std::cout << "Accuracy: train = " << trainAccuracy << "%,"
            << "\t valid = " << validAccuracy << "%" << std::endl;

  mlpack::data::Save("model.bin", "model", model, false);

  std::cout << "Predicting ..." << std::endl;

  FFN<> model2;
  mlpack::data::Load("model.bin", "model", model2);
  // Getting predictions on test data points .
  model2.Predict(validX, predOut);
  // Calculating accuracy on validating data points.
  predLabels = getLabels(predOut);
  validAccuracy =
      arma::accu(predLabels == validY) / (double) validY.n_elem * 100;

  std::cout << "Accuracy: " << " valid = " << validAccuracy << "%" << endl;
}

```

Or you can clone this [branch](https://github.com/kartikdutt18/examples/tree/SerializationError) and run `make` in mnist_cnn directory.

The file above uses the saved model to generate predictions. The results should be same of the runtime model and the saved model however they differ.
Following are the last few lines of the output.

```
Accuracy: train = 88.737%,       valid = 88.4481%
Predicting ...
Accuracy:  valid = 19.7533%
```

**Comment by zoq** [comment]
Not sure maybe you already tested the code above without the `BatchNorm` layer? Does it produce the same result?

**Comment by kartikdutt18** [comment]
I hadn't earlier but this problem isn't there without BatchNorm Layer. So the problem is with serialization of BatchNorm Layer.

**Comment by saksham189** [comment]
can you try to compare the parameters for model and model2 and identify where the difference is?

**Comment by kartikdutt18** [comment]
Sure, that makes sense. I will test it out.

**Comment by zoq** [comment]
Looking at the forward pass:

https://github.com/mlpack/mlpack/blob/da83453e24b1e7bb704e8174a912af48d5d1c2d7/src/mlpack/methods/ann/layer/batch_norm_impl.hpp#L154-L165

each parameter should be serialized, but still a good idea to check if the parameter are correctly loaded.

**Comment by rcurtin** [comment]
Just glancing at the implementation, I looked through members that aren't serialized; it led to a few questions about the implementation:

 1. It seems like maybe `mean` and `variance` need to be serialized.
 2. We seem to be serializing `gamma` and `beta`, but `Reset()` sets those to aliases of the `weights` matrix; so I wonder if we can avoid serializing those.
 3. Maybe `normalized` needs to be serialized; I'm not sure.

Anyway, maybe those are some helpful thoughts.  To me it seems like `mean` and `variance` are the main candidates for the issue.
